lib/dRep/dRepImpl.py

```diff
@@ -80,102 +77,321 @@ def run_dRep(self, ctx, params):
-        }
+            dRep_cmd = ("Skipped running dRep ... "
+                        "but dRep_params are: "
+                        f"`{' '.join(dRep_params)}` and workDir is `{dRep_workDir}`")
+
+        else:
+            dRep_workDir = os.path.join(_globals.run_dir, 'dRep_workDir')
+
+            dRep_cmd = ' '.join(
+                [f'dRep dereplicate {dRep_workDir} --genomes {binsPooled_dir}/*.fasta'] + dRep_params)
+
+            logging.info('Running dRep command `%s`' % dRep_cmd)
+            completed_proc = subprocess.run(dRep_cmd, shell=True, stdout=sys.stdout, stderr=sys.stderr)
+
+            retcode = completed_proc.returncode
+
+            dprint('retcode', run=locals())
+
+            if retcode != 0:
+
+                # check: if Bdb.csv is empty -> nothing passed length/qual filtering
+                with open(os.path.join(dRep_workDir, 'data_tables/Bdb.csv')) as f:
+                    num_lines = sum(1 for line in f)
+
+                if num_lines == 1:
+                    msg = 'Sorry, dRep terminated with return code %d because no bins passed length and CheckM filtering' % retcode
+
+                else:
+                    msg = (
+                        "Sorry, commmand: `%s` terminated abnormally with return code: %d"
+                        % dRep_cmd, retcode
+                        )
+
+                raise Exception(msg)
 
-        dRep_binnedContigs_objData = self.mgu.file_to_binned_contigs(mguFileToBinnedContigs_params)
 
-        {
-            'obj_name': dRep_binnedContigs_objName,
-            'obj_ref': dRep_binnedContigs_objData['binned_contig_obj_ref']
-        }
-       
-        objects_created = [{'ref': dRep_binnedContigs_objData['binned_contig_obj_ref']}]
+
+
+
+        #
+        ##
+        ### result BinnedContigs
+        ####
+        #####
+
+        bins_derep_dir = os.path.join(dRep_workDir, 'dereplicated_genomes')
+        objects_created = []
+
+        # for each original BinnedContigs
+        for bc in BinnedContigs.created_instances:
+            logging.info(f'Dereplicating {bc.name}')
+            bc.reduce_to_dereplicated(bins_derep_dir)
+
+            if not bc.is_empty():
+                if _globals.debug and params.get('skip_save_bc'):
+                    BinnedContigs.saved_instances.append(bc)
+                else: 
+                    objects_created.append(bc.save())
+            
+
+
+
+        #
+        ##
+        ### html
+        ####
+        #####
+
+        html_dir = os.path.join(_globals.run_dir, 'html_dir')
+        shutil.copytree('/kb/module/ui/output', html_dir) # dir of html and accessories
+
+        report.HTMLBuilder(BinnedContigs.saved_instances, dRep_cmd, dRep_workDir, html_dir)
+
 
 
         #
         ##
-        ### Pdf outputs
+        ### to shock
         ####
         #####
 
+        if _globals.debug and params.get('skip_save_shock'):
+            return
+
+
+        def dir_to_shock(dir_path, name, description):
```

Comment:
Are we still in the dereplicate method? I'm losing track. It'd be cleaner to move this function into a separate module and import it.